src/main/java/net/glowstone/net/handler/status/StatusRequestHandler.java

```diff
@@ -1,79 +1,95 @@
 package net.glowstone.net.handler.status;
 
+import com.destroystokyo.paper.event.server.PaperServerListPingEvent;
+import com.destroystokyo.paper.network.StatusClient;
+import com.destroystokyo.paper.profile.PlayerProfile;
 import com.flowpowered.network.MessageHandler;
-import java.net.InetAddress;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Strings;
 import java.util.ArrayList;
 import java.util.Collections;
-import java.util.Iterator;
 import java.util.List;
+import java.util.UUID;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
 import net.glowstone.EventFactory;
 import net.glowstone.GlowServer;
+import net.glowstone.entity.GlowPlayer;
 import net.glowstone.net.GlowSession;
+import net.glowstone.net.GlowStatusClient;
 import net.glowstone.net.message.status.StatusRequestMessage;
 import net.glowstone.net.message.status.StatusResponseMessage;
-import net.glowstone.util.GlowServerIcon;
 import org.bukkit.entity.Player;
-import org.bukkit.event.server.ServerListPingEvent;
 import org.bukkit.util.CachedServerIcon;
 import org.json.simple.JSONArray;
 import org.json.simple.JSONObject;
 
 public final class StatusRequestHandler implements
     MessageHandler<GlowSession, StatusRequestMessage> {
 
+    private static final UUID FAKE_UUID = new UUID(0, 0);
+
     @Override
     @SuppressWarnings("unchecked")
     public void handle(GlowSession session, StatusRequestMessage message) {
         // create and call the event
         GlowServer server = session.getServer();
-        int online = server.getOnlinePlayers().size();
-        InetAddress address = session.getAddress().getAddress();
+        StatusEvent event = new StatusEvent(
+                new GlowStatusClient(session), server.getMotd(),
+                server.getOnlinePlayers().size(), server.getMaxPlayers(),
+                GlowServer.GAME_VERSION, GlowServer.PROTOCOL_VERSION,
+                server.getServerIcon());
 
-        StatusEvent event = new StatusEvent(address, server.getMotd(), online,
-            server.getMaxPlayers());
-        event.players = new ArrayList<>(server.getOnlinePlayers());
-        event.icon = server.getServerIcon();
         event.serverType = server.getServerType();
         event.clientModsAllowed = server.getAllowClientMods();
+
+        // Add player sample list
+        List<Player> playerList = new ArrayList<>(server.getOnlinePlayers());
+        int count = Math.min(playerList.size(), server.getPlayerSampleCount());
+        Collections.shuffle(playerList);
```

Comment:
Is shuffling really necessary? I understand if that's Vanilla behavior, but it seems like to me like an unnecessary performance hit (especially with a large player count)

Reply:
The main purpose here is to avoid showing the same players all the time. This could be simplified more, but then it would be less random. Up to you.

Reply:
Implement a shuffle that stops choosing players when you have count.

Reply:
What does "when you have count" mean?

Reply:
I mean, if more than `server.getPlayerSampleCount() > playerList.size()`, then only choose `playerList.size()` random elements, and stop building the shuffled list once you have that many.